### PR TITLE
Fix support for logging within callbacks returned from `LightningModule`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

The logging support for callbacks is added before the callbacks from LightningModule are configured. The PR fixes that. 
~Also added a small support of returning a single callback from `LightningModule.configure_callbacks` rather than a list always.~

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

Part of #1

cc @borda @awaelchli @ananthsub @daniellepintz @carmocca @edward-io

### CODE REVIEW

File: tests/trainer/connectors/test_callback_connector.py [hunk 1]
```diff
@@ -211,6 +211,22 @@ def test_attach_model_callbacks_override_info(caplog):
...
+
+def test_attach_model_callbacks_support_logging(tmpdir):
+    class TempCallback(Callback):
+        def on_train_epoch_start(self, *args, **kwargs):
+            self.log("on_train_epoch_end", 7.0)
+
+    class CustomBoringModel(BoringModel):
+        def configure_callbacks(self):
+            cb = TempCallback()
+            return cb
```
Comment (tchaton) [inline tests/trainer/connectors/test_callback_connector.py:227]:
```suggestion
            return TempCallback()
```

**Review by tchaton — APPROVED** [review]
Great catch !

**Comment by awaelchli** [comment]
> Also added a small support of returning a single callback from LightningModule.configure_callbacks rather than a list always.

This would have to be a separate PR. We want to include only the fix in the bugfix release.

File: tests/trainer/connectors/test_callback_connector.py [hunk 2]
```diff
@@ -211,6 +211,21 @@ def test_attach_model_callbacks_override_info(caplog):
     trainer.model = model
     cb_connector = CallbackConnector(trainer)
     with caplog.at_level(logging.INFO):
-        cb_connector._attach_model_callbacks()
+        cb_connector._attach_model_callbacks(model)
 
     assert "existing callbacks passed to Trainer: EarlyStopping, LearningRateMonitor" in caplog.text
+
+
+def test_attach_model_callbacks_support_logging(tmpdir):
```
Comment (awaelchli) [inline tests/trainer/connectors/test_callback_connector.py:219]: I believe the test_hooks.py is sufficient and no extra test needs to be added.

Comment (rohitgr7) [reply tests/trainer/connectors/test_callback_connector.py:219]: yeah but no tests were failing, right?

Comment (rohitgr7) [reply tests/trainer/connectors/test_callback_connector.py:219]: oh ! okay. got it.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1091,6 +1091,7 @@ def _run(
         if hasattr(model, "hparams"):
             parsing.clean_namespace(model.hparams)
 
+        self._callback_connector._attach_model_callbacks(model)
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1094]: why before ""verify_loop_configurations" and not under `attach_model_logging_functions`? That way you don't need to pass the pl_module reference as we want to remove the argument

cc @daniellepintz

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1094]: under `verify_loop_configurations` we have certain checks for callback hooks. So here now we make sure we check them for callbacks returned from `LightningModule` as well.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py]: Pushed a change reordering so the argument change is not necessary.

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -1091,17 +1091,16 @@ def _run(
...
-        # attach model log function to callback
-        self._callback_connector.attach_model_logging_functions(model)
-
         # attach model to the training type plugin
         self.training_type_plugin.connect(model)
 
+        self._callback_connector._attach_model_callbacks()
+        self._callback_connector._attach_model_logging_functions()
+
+        verify_loop_configurations(self, model)
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:1100]: thanks, @carmocca,
I guess with the changes you made we can remove the model from `verify_loop_configurations` as well?
